File: duplicates-android/app/src/main/java/com/github/duplicates/calendar/CalendarComparator.java
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.calendar;

import com.android.calendarcommon2.RecurrenceSet;
import com.github.duplicates.DuplicateComparator;

import java.util.Calendar;
import java.util.TimeZone;

import static android.text.format.DateUtils.MINUTE_IN_MILLIS;
import static com.android.calendarcommon2.EventRecurrence.DAILY;
import static com.android.calendarcommon2.EventRecurrence.HOURLY;
import static com.android.calendarcommon2.EventRecurrence.MINUTELY;
import static com.android.calendarcommon2.EventRecurrence.MONTHLY;
import static com.android.calendarcommon2.EventRecurrence.SECONDLY;
import static com.android.calendarcommon2.EventRecurrence.WEEKLY;
import static com.android.calendarcommon2.EventRecurrence.YEARLY;
import static com.github.duplicates.calendar.CalendarItem.NEVER;

/**
 * Compare duplicate calendar events.
 *
 * @author moshe.w
 */
public class CalendarComparator extends DuplicateComparator<CalendarItem> {

    public static final int TITLE = 0;
    public static final int DTSTART = 1;
    public static final int DTEND = 2;
    public static final int DESCRIPTION = 3;
    public static final int LOCATION = 4;
    public static final int ATTENDEES = 5;

    private final Calendar cal1 = Calendar.getInstance();
    private final Calendar cal2 = Calendar.getInstance();

    @Override
    public int compare(CalendarItem lhs, CalendarItem rhs) {
        int c;

        c = compareIgnoreCase(lhs.getTitle(), rhs.getTitle());
        if (c != SAME) {
            return c;
        }
        c = compareIgnoreCase(lhs.getDescription(), rhs.getDescription());
        if (c != SAME) {
            return c;
        }
        c = compareIgnoreCase(lhs.getLocation(), rhs.getLocation());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getStart(), rhs.getStart());
        if (c != SAME) {
            return c;
        }
        c = compare(lhs.getEndEffective(), rhs.getEndEffective());
        if (c != SAME) {
            return c;
        }

        return super.compare(lhs, rhs);
    }

    @Override
    public boolean[] difference(CalendarItem lhs, CalendarItem rhs) {
        boolean[] result = new boolean[6];

        result[ATTENDEES] = compare(lhs.isHasAttendeeData(), rhs.isHasAttendeeData()) != SAME;
        result[DESCRIPTION] = compare(lhs.getDescription(), rhs.getDescription()) != SAME;
        result[DTSTART] = compare(lhs.getStart(), rhs.getStart()) != SAME;
        result[DTEND] = compare(lhs.getEndEffective(), rhs.getEndEffective()) != SAME;
        result[LOCATION] = compare(lhs.getLocation(), rhs.getLocation()) != SAME;
        result[TITLE] = compare(lhs.getTitle().toLowerCase(), rhs.getTitle().toLowerCase()) != SAME;

        return result;
    }

    @Override
    public float match(CalendarItem lhs, CalendarItem rhs, boolean[] difference) {
        if (difference == null) {
            difference = difference(lhs, rhs);
        }
        float match = MATCH_SAME;

        if (difference[TITLE]) {
            match *= 0.85f;
        }
        if (difference[DTSTART]) {
            match *= matchDate(lhs.getStart(), lhs.getStartTimeZone(), lhs.getRecurrenceSet(), lhs.isAllDay(), rhs.getStart(), rhs.getStartTimeZone(), rhs.getRecurrenceSet(), rhs.isAllDay());
        }
        if (difference[DTEND]) {
            match *= matchDate(lhs.getEndEffective(), lhs.getEndTimeZone(), lhs.getRecurrenceSet(), lhs.isAllDay(), rhs.getEndEffective(), rhs.getEndTimeZone(), rhs.getRecurrenceSet(), rhs.isAllDay());
        }

        if (difference[DESCRIPTION]) {
            match *= 0.95f;
        }
        if (difference[LOCATION]) {
            match *= 0.96f;
        }
        if (difference[ATTENDEES]) {
            match *= 0.96f;
        }

        return match;
    }

    protected float matchDate(long lhs, TimeZone lhsTimeZone, RecurrenceSet lhsRecurrence, boolean lhsAllDay, long rhs, TimeZone rhsTimeZone, RecurrenceSet rhsRecurrence, boolean rhsAllDay) {
        long dt = Math.abs(lhs - rhs);
        if (dt < MINUTE_IN_MILLIS) {
            return MATCH_SAME;
        }

        if ((lhsRecurrence.rrules != null) && (rhsRecurrence.rrules != null)) {
            cal1.setTimeZone(lhsTimeZone);
            cal1.setTimeInMillis(lhs);
            cal2.setTimeZone(rhsTimeZone);
            cal2.setTimeInMillis(rhs);

            if (lhsRecurrence.rrules[0].freq == rhsRecurrence.rrules[0].freq) {
                if ((lhs == NEVER) || (rhs == NEVER)) {
                    return 0.95f;
                }

                boolean same = false;

                switch (lhsRecurrence.rrules[0].freq) {
                    case SECONDLY:
                        same = isSame(cal1, cal2, Calendar.MILLISECOND);
                        break;
                    case MINUTELY:
                        same = isSame(cal1, cal2, Calendar.SECOND);
                        break;
                    case HOURLY:
                        same = isSame(cal1, cal2, Calendar.MINUTE);
                        break;
                    case DAILY:
                        same = isSame(cal1, cal2, Calendar.HOUR_OF_DAY, Calendar.MINUTE);
                        break;
                    case WEEKLY:
                        same = isSame(cal1, cal2, Calendar.DAY_OF_WEEK, Calendar.HOUR_OF_DAY);
                        break;
                    case MONTHLY:
                        same = isSame(cal1, cal2, Calendar.DAY_OF_MONTH, Calendar.HOUR_OF_DAY);
                        break;
                    case YEARLY:
                        if (lhsAllDay && rhsAllDay) {
                            same = isSame(cal1, cal2, Calendar.MONTH, Calendar.DAY_OF_MONTH);
                        } else {
                            same = isSame(cal1, cal2, Calendar.MONTH, Calendar.DAY_OF_MONTH, Calendar.HOUR_OF_DAY);
                        }
                        break;
                }
                if (same) {
                    return MATCH_SAME;
                }
            }
        }

        return 0.8f;
    }

    protected boolean isSame(Calendar lhs, Calendar rhs, int... fields) {
        long dt = Math.abs(lhs.getTimeInMillis() - rhs.getTimeInMillis());
        if (dt < MINUTE_IN_MILLIS) {
            return true;
        }

        boolean same = true;
        for (int field : fields) {
            same &= lhs.get(field) == rhs.get(field);
        }

        return same;
    }
}

File: duplicates-android/app/src/main/java/com/github/duplicates/calendar/CalendarComparator.kt
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.calendar

import android.text.format.DateUtils.MINUTE_IN_MILLIS
import com.android.calendarcommon2.EventRecurrence.DAILY
import com.android.calendarcommon2.EventRecurrence.HOURLY
import com.android.calendarcommon2.EventRecurrence.MINUTELY
import com.android.calendarcommon2.EventRecurrence.MONTHLY
import com.android.calendarcommon2.EventRecurrence.SECONDLY
import com.android.calendarcommon2.EventRecurrence.WEEKLY
import com.android.calendarcommon2.EventRecurrence.YEARLY
import com.android.calendarcommon2.RecurrenceSet
import com.github.duplicates.DuplicateComparator
import com.github.duplicates.calendar.CalendarItem.Companion.NEVER
import java.util.*
import java.util.Calendar.DAY_OF_MONTH
import java.util.Calendar.DAY_OF_WEEK
import java.util.Calendar.HOUR_OF_DAY
import java.util.Calendar.MILLISECOND
import java.util.Calendar.MINUTE
import java.util.Calendar.MONTH
import java.util.Calendar.SECOND

/**
 * Compare duplicate calendar events.
 *
 * @author moshe.w
 */
class CalendarComparator : DuplicateComparator<CalendarItem>() {

    private val cal1 = Calendar.getInstance()
    private val cal2 = Calendar.getInstance()

    override fun compare(lhs: CalendarItem, rhs: CalendarItem): Int {
        var c: Int

        c = compareIgnoreCase(lhs.title, rhs.title)
        if (c != SAME) return c
        c = compareIgnoreCase(lhs.description, rhs.description)
        if (c != SAME) return c
        c = compareIgnoreCase(lhs.location, rhs.location)
        if (c != SAME) return c
        c = compare(lhs.start, rhs.start)
        if (c != SAME) return c
        c = compare(lhs.endEffective, rhs.endEffective)
        return if (c != SAME) c else super.compare(lhs, rhs)
    }

    override fun difference(lhs: CalendarItem, rhs: CalendarItem): BooleanArray {
        val result = BooleanArray(6)
        if (lhs.id == rhs.id) {
            return result
        }

        result[ATTENDEES] = isDifferent(lhs.isHasAttendeeData, rhs.isHasAttendeeData)
        result[DESCRIPTION] = isDifferent(lhs.description, rhs.description)
        result[DTSTART] = isDifferent(lhs.start, rhs.start)
        result[DTEND] = isDifferent(lhs.endEffective, rhs.endEffective)
        result[LOCATION] = isDifferent(lhs.location, rhs.location)
        result[TITLE] = isDifferent(lhs.title.toLowerCase(locale), rhs.title.toLowerCase(locale))

        return result
    }

    override fun match(lhs: CalendarItem, rhs: CalendarItem, difference: BooleanArray?): Float {
        val different = difference ?: difference(lhs, rhs)
        var match = MATCH_SAME
        if (lhs.id == rhs.id) {
            return 0f
        }

        if (different[TITLE]) {
            match *= matchTitle(lhs.title, rhs.title, 0.85f)
        }
        if (different[DTSTART]) {
            match *= matchDate(
                lhs.start, lhs.getStartTimeZoneNN(), lhs.recurrenceSet, lhs.isAllDay,
                rhs.start, rhs.getStartTimeZoneNN(), rhs.recurrenceSet, rhs.isAllDay
            )
        }
        if (different[DTEND]) {
            match *= matchDate(
                lhs.endEffective, lhs.getEndTimeZoneNN(), lhs.recurrenceSet, lhs.isAllDay,
                rhs.endEffective, rhs.getEndTimeZoneNN(), rhs.recurrenceSet, rhs.isAllDay
            )
        }
        if (different[DESCRIPTION]) {
            match *= 0.95f
        }
        if (different[LOCATION]) {
            match *= 0.96f
        }
        if (different[ATTENDEES]) {
            match *= 0.96f
        }

        return match
    }

    protected fun matchDate(
        lhs: Long,
        lhsTimeZone: TimeZone,
        lhsRecurrence: RecurrenceSet,
        lhsAllDay: Boolean,
        rhs: Long,
        rhsTimeZone: TimeZone,
        rhsRecurrence: RecurrenceSet,
        rhsAllDay: Boolean
    ): Float {
        cal1.timeZone = lhsTimeZone
        cal1.timeInMillis = lhs
        cal2.timeZone = rhsTimeZone
        cal2.timeInMillis = rhs
        val dt = Math.abs(cal1.timeInMillis - cal2.timeInMillis)
        if (dt < MINUTE_IN_MILLIS) {
            return MATCH_SAME
        }

        if ((lhsRecurrence.rrules != null) && (rhsRecurrence.rrules != null)) {
            if (lhsRecurrence.rrules[0].freq == rhsRecurrence.rrules[0].freq) {
                if ((lhs == NEVER) || (rhs == NEVER)) {
                    return 0.95f
                }

                var same = false

                when (lhsRecurrence.rrules[0].freq) {
                    SECONDLY -> same = isSame(cal1, cal2, MILLISECOND)
                    MINUTELY -> same = isSame(cal1, cal2, SECOND)
                    HOURLY -> same = isSame(cal1, cal2, MINUTE)
                    DAILY -> same = isSame(cal1, cal2, HOUR_OF_DAY, MINUTE)
                    WEEKLY -> same = isSame(cal1, cal2, DAY_OF_WEEK, HOUR_OF_DAY)
                    MONTHLY -> same = isSame(cal1, cal2, DAY_OF_MONTH, HOUR_OF_DAY)
                    YEARLY -> if (lhsAllDay && rhsAllDay) {
                        same = isSame(cal1, cal2, MONTH, DAY_OF_MONTH)
                    } else {
                        same = isSame(cal1, cal2, MONTH, DAY_OF_MONTH, HOUR_OF_DAY)
                    }
                }
                if (same) {
                    return 0.99f
                }
            }
        }

        return 0.8f
    }

    protected fun isSame(lhs: Calendar, rhs: Calendar, vararg fields: Int): Boolean {
        var same = true
        for (field in fields) {
            same = same and (lhs.get(field) == rhs.get(field))
        }
        return same
    }

    companion object {

        const val TITLE = 0
        const val DTSTART = 1
        const val DTEND = 2
        const val DESCRIPTION = 3
        const val LOCATION = 4
        const val ATTENDEES = 5
    }
}
